src/shared/pkcs11-util.c

```diff
@@ -623,105 +630,389 @@ int pkcs11_token_find_private_key(
+                rv = m->C_GetAttributeValue(session, candidate, optional_attributes, ELEMENTSOF(optional_attributes));
+                if (rv != CKR_OK && rv != CKR_ATTRIBUTE_TYPE_INVALID)
+                        return log_error_errno(SYNTHETIC_ERRNO(EIO),
+                                "Failed to get attributes of a selected private key: %s", sym_p11_kit_strerror(rv));
+
+                if (optional_attributes[0].ulValueLen != CK_UNAVAILABLE_INFORMATION && decrypt_value == CK_TRUE)
+                        can_decrypt = true;
+
+                if (optional_attributes[1].ulValueLen != CK_UNAVAILABLE_INFORMATION && derive_value == CK_TRUE)
+                        can_derive = true;
+
+                if (can_decrypt || can_derive) {
+                        if (n_objects == 0) {
+                                object = candidate;
+                        }
+                        n_objects++;
+                        if (n_objects > 1)
+                                break;
                 }
+        }
 
-                case CKA_KEY_TYPE: {
-                        CK_KEY_TYPE t;
+        rv = m->C_FindObjectsFinal(session);
+        if (rv != CKR_OK)
+                return log_error_errno(SYNTHETIC_ERRNO(EIO),
+                        "Failed to finalize object find call: %s", sym_p11_kit_strerror(rv));
 
-                        if (attributes[a].ulValueLen != sizeof(t))
-                                return log_error_errno(SYNTHETIC_ERRNO(EINVAL), "Invalid PKCS#11 CKA_KEY_TYPE attribute size.");
+        if (n_objects == 0)
+                return log_error_errno(SYNTHETIC_ERRNO(ENOENT),
+                        "Failed to find selected private key suitable for decryption or derivation on token.");
 
-                        memcpy(&t, attributes[a].pValue, sizeof(t));
-                        if (t != CKK_RSA)
-                                return log_error_errno(SYNTHETIC_ERRNO(EINVAL), "Selected PKCS#11 object is not an RSA key, refusing.");
+        if (n_objects > 1)
+                return log_error_errno(SYNTHETIC_ERRNO(ENOTUNIQ),
+                        "Configured private key URI matches multiple keys, refusing.");
 
-                        found_key_type = true;
-                        break;
-                }}
+        *ret_object = object;
+        return 0;
+}
+
+static const char* object_class_to_string(CK_OBJECT_CLASS class) {
+        switch (class) {
+        case CKO_CERTIFICATE: return "CKO_CERTIFICATE";
+        case CKO_PUBLIC_KEY:  return "CKO_PUBLIC_KEY";
+        case CKO_PRIVATE_KEY: return "CKO_PRIVATE_KEY";
+        case CKO_SECRET_KEY:  return "CKO_SECRET_KEY";
+        default:              return "Unknown";
         }
+}
 
-        if (!found_decrypt || !found_class || !found_key_type) {
-                /* Hmm, let's slightly extend the attribute list we search for */
+// Returns an object with the given class and the same CKA_ID or CKA_LABEL as prototype
```

Comment:
as per coding style: we use // style comments only for "fixme" comments, and /* */ style comments for explanatory stuff. hence please switch to /* */ here